docs/RFCS/20222202_historical_hot_ranges.md

```diff
@@ -0,0 +1,224 @@
+
+
+<br/><br/>
+
+A new function will implement this behavior:
+```go
+type ReportedRange struct {
+    qps float32
+    startKey string
+    endkey string
+}
+
+type AggregatedRanges struct {
+    storeId int
+    timestamp int
+    ranges []ReportedRange
+}
+
+func (s *Store) HottestRangesAggregated(
+    nStores int,
+    lastAvgHottness float, 
+    lastAvgHottnessLocal float
+) AggregatedRanges
+```
+Open Questions:
+1. What happens if the budget is > 128? This is relevant for small clusters with few ranges.
+
+<br/><br/>
+The functionality described in this section is pictured here:
+
+![HottestRangesAggregated](./images/20222202_historical_hot_ranges/range-aggregation.svg "HottestRangesAggregated")
+
+_____
+
+### Querying
+In order to render the visualization, the client must request historical data for the time period of interest.
+285 Megabytes is far too much to send over the wire in a single payload, so requests for historical data must be broken into discrete chunks. The number of requests made depends on the timespan the client wants to visualize, with the default being 6 hours, or 24 samples of HHR per chunk.
+
+Rendering the visualization is not dependent on the majority of bytes associated within a sample. To render a sample, the browser only needs to know:
+
+1. The timestamp
+2. The QPS values for ranges in the sample
+3. The start_key for ranges in the sample
+
+The `start_key` is required because the visualization needs to maintain the set of all keys. (The requirements for the visualization itself is discussed later on.)
+
+When a user wants to know information about a hot range bucket, they will hover their cursor over a particular bucket. A debounced request will be made to the server requesting the metadata for the bucket of interest. Because we have the range's start key, we can use that and the timestamp to query `crdb_internal.historical_hot_ranges`, allowing us to save bandwidth by not needing to send the uuid over the wire for each range for each sample. This comes at the cost of increased storage, because we have to maintain an index on the `start_key`.
+
+
+Open Questions:
+1. Would it suffice to send the first 10 or 20 characters of the `start_key`, trading resolution for performance? (Assuming we'd send the uuid too.)
+
+<br/><br/>
+The full querying process is pictured here:
+![Querying](./images/20222202_historical_hot_ranges/requests.svg "Querying")
+
+
+____
+### Visualization
+
+The Key Visualizer has the following features:
+- Time is represented by the x-axis
+- Key space is represented by the y-axis
+- Hotness is represented by a spectrum of color, with deep blue representing QPS activity equal to zero, and White equal to the max QPS activity seen across the time period being visualized.
```

Comment:
Range size is very important to hotness in addition to QPS. Access density is sometimes used as a metric such as QPS per unit of storage.

Reply:
Ack.